### ISSUE TITLE
Suggestion: Alternative way to treat outdated ini files

### ISSUE THREAD

**Issue by bernt-matthias** [OpenMS/OpenMS#4462]
Clearly its a good idea to have the ability to run tools using old ini files. But maybe it would be better to have a single tool that tries to update an ini file to the current state, e.g. `UpdateIniFile -in old.ini -out current.ini`, which then can be used in the most recent version of the tools. The tools itself might just raise an error if the ini version is not up-to-date. 

With the current state I see the problem that there seems to be no warning / error if the mapping of the parameters in the ini file could be mapped (besides the warning that an outdated ini file was used). Hence the user does not know if / which parameters are not correctly set. And many might just ignore it.

Related

https://github.com/OpenMS/OpenMS/issues/4386
https://github.com/OpenMS/OpenMS/issues/4461

**Comment by jpfeuffer** [comment]
We have the INIUpdater util. Is that what you are looking for?

But I am not familiar right now with the warnings that are given when reading an ini.
Ideally it should warn about unused and unspecified parameters, I agree.

**Comment by bernt-matthias** [comment]
Sounds exactly like what could help me. Unfortunately it seems the the binary is not included in the conda package. 

Is there a specific make target that would build it?

(Just have seen your recent work on the bioconda recipe .. cool)

**Comment by jpfeuffer** [comment]
Yes, "unfortunately" INIUpdater also has a GUI. On conda we currently build without GUI since it would include a lot more dependencies (see here https://github.com/ContinuumIO/anaconda-issues/issues/8779)

Doesn't INIUpdater have its own target? Try "make INIUpdater".

**Comment by cbielow** [comment]
Not sure if this everything you need, but from what I see, this functionality already exists:

```
FileFilter.exe -write_ini c:\temp\ff.ini
```
Now I edit the file, i.e. remove the 'log' param and rename another one, plus I downgrade the version (to simulate an older ini file)

Then running the tool with that old file:
```
FileFilter.exe -ini c:\temp\ff.ini -in "c:\temp\170916 W3 A P60 KNW PNA 1_36.mzML" -out c:\temp\bla.mzML
Unknown (or deprecated) Parameter 'peak_options:completly_unknown_option' given in outdated parameter file!
Parameters passed to 'FileFilter' are invalid. To prevent usage of wrong defaults, please update/fix the parameters!
```
.. so this works as expected.
The only case where TOPP tools accept outdated parameters is when they just moved locations, but keep their name (see other tickets).

To fix this, simply combine `-ini` + `-write_ini`
```
FileFilter.exe -ini c:\temp\ff.ini -write_ini c:\temp\ff_new.ini
Warning: for ':version' entry, augmented and Default Ini-File differ in value. Default value will not be altered!
Unknown (or deprecated) Parameter 'FileFilter:1:peak_options:completly_unknown_option' given in outdated parameter file! Ignoring parameter.
```

and so, the ini is fixed...

```
\FileFilter.exe -ini c:\temp\ff_new.ini -in "c:\temp\170916 W3 A P60 KNW PNA 1_36.mzML" -out c:\temp\bla.mzML
Progress of 'loading chromatogram list':
-- done [took 0.03 s (CPU), 0.02 s (Wall)] --
Progress of 'storing mzML file':
-- done [took 0.00 s (CPU), 0.01 s (Wall)] --
FileFilter took 0.49 s (wall), 0.47 s (CPU), 0.16 s (system), 0.31 s (user).
```

**Comment by cbielow** [comment]
This is 'documented' in the CHANGELOG, but should probably also be mentioned in `TOPP_scripting.doxygen`

**Comment by bernt-matthias** [comment]
So I could run the tests and add `-write_ini` if `-ini` is used an have up to date ini files afterwards?

**Comment by cbielow** [comment]
if you combine `-write_ini` with `-ini` all the tool will do is update the file and quit (no computation will be done). If fact, `-write_ini` alone will do the same (if I remember correctly).
So you'd need to call the tool twice (once for update, second time for running with the new ini)

**Comment by bernt-matthias** [comment]
I was just trying this approach and found that (unfortunately) it may change the values contained in the ini file. 

For instance 
`MascotAdapter -ini MascotAdapter_1_parameters.ini -write_ini MascotAdapter_1_parameters.ini`

changes the output file and many parameters. This ini file seems to be quite old, .e.g. the output file is a string and not output-file... 

So far this is the first instance. But I did only look into very few (i.e. 2) examples.
